source/server/filter_chain_manager_impl.cc

```diff
@@ -496,6 +571,19 @@ FilterChainManagerImpl::findFilterChain(const Network::ConnectionSocket& socket)
              : default_filter_chain_.get();
 }
 
+const Network::FilterChain*
+FilterChainManagerImpl::findFilterChainUsingMatcher(const Network::ConnectionSocket& socket) const {
+  Network::Matching::MatchingDataImpl data(socket);
+  const auto& match_result = Matcher::evaluateMatch<Network::MatchingData>(*matcher_, data);
+  ASSERT(match_result.match_state_ == Matcher::MatchState::MatchComplete,
+         "Matching must complete for network streams.");
+  if (match_result.result_) {
+    const auto result = match_result.result_();
+    return result->getTyped<FilterChainNameAction>().chain_.get();
+  }
+  return default_filter_chain_.get();
```

Comment:
Maybe add a trace/debug log here per @howardjohn's point on debugging?

Reply:
Added a debug log when matcher refers to a missing chain. I think that's the confusing state that @howardjohn referred to.